ibtk/src/lagrangian/BoxPartitioner.cpp

```diff
@@ -0,0 +1,332 @@
+        position.resize(position_solution->size());
+        position_solution->localize(position);
+    }
+    std::vector<libMesh::Point> node_positions(mesh.parallel_n_nodes());
+    std::size_t node_n = 0;
+    for (auto node_it = mesh.nodes_begin(); node_it != mesh.nodes_end(); ++node_it)
+    {
+        const libMesh::Node* const node = *node_it;
+        TBOX_ASSERT(node->id() == node_n);
+        libMesh::Point node_position;
+        if (use_position_vector)
+        {
+            if (node->n_vars(position_system_n))
+            {
+                TBOX_ASSERT(node->n_vars(position_system_n) == NDIM);
+                for (unsigned int d = 0; d < NDIM; ++d)
+                {
+                    node_position(d)
+                        = position[node->dof_number(position_system_n, d, 0)];
+                }
+            }
+        }
+        else
+        {
+            node_position = *node;
+        }
+        node_positions[node_n] = node_position;
+        ++node_n;
+    }
+    TBOX_ASSERT(node_n == node_positions.size());
+
+    // Step 1: determine which elements belong to which processor and
+    // communicate the partitioning information across the network.
+    //
+    // We offset the processor ids by 1 so that we can check, by summation,
+    // that each Elem and Node is given exactly one processor id.
+    std::vector<libMesh::processor_id_type> elem_ids(mesh.parallel_n_elem());
+    std::vector<libMesh::dof_id_type> local_elem_ids;
+    std::size_t elem_n = 0;
+    const auto end_elem = mesh.active_elements_end();
+    for (auto elem = mesh.active_elements_begin(); elem != end_elem; ++elem)
+    {
+        libMesh::Point centroid;
+        const unsigned int n_nodes = (*elem)->n_nodes();
+        for (unsigned int node_n = 0; node_n < n_nodes; ++node_n)
+        {
+            centroid += node_positions[(*elem)->node_id(node_n)];
+        }
+        centroid *= 1.0/n_nodes;
+
+        if (d_partitioning_boxes.contains(to_ibtk_point(centroid)))
+        {
+            local_elem_ids.push_back(elem_n);
+            TBOX_ASSERT(elem_n < elem_ids.size());
+            elem_ids[elem_n] = current_rank + 1;
+        }
+        ++elem_n;
+    }
+
+    std::vector<libMesh::processor_id_type> node_ids(mesh.parallel_n_nodes());
+    std::vector<libMesh::dof_id_type> local_node_ids;
+    std::size_t n_local_nodes = 0;
+    node_n = 0;
+    const auto end_node = mesh.active_nodes_end();
+    for (auto node = mesh.active_nodes_begin(); node != end_node; ++node)
+    {
+        if (d_partitioning_boxes.contains(to_ibtk_point(node_positions[node_n])))
+        {
+            local_node_ids.push_back(node_n);
+            TBOX_ASSERT(node_n < node_ids.size());
+            node_ids[node_n] = current_rank + 1;
+            ++n_local_nodes;
+        }
+        ++node_n;
+    }
+
+    // step 1.5: log the current state of affairs:
+    const int n_processes = SAMRAI::tbox::SAMRAI_MPI::getNodes();
```

Comment:
Should we make this optional, and off by default?

Reply:
Sure.

(note to self: add `bool d_enable_logging = false;` to the header and initialize it in the constructor.)

Reply:
Fixed!